validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java

```diff
@@ -0,0 +1,284 @@
+    List<SignedBeaconBlock> producedBlocks =
+        producedBlocksByEpoch.subMap(startEpochInclusive, true, endEpochExclusive, false).values()
+            .stream()
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+
+    long numberOfIncludedBlocks =
+        producedBlocks.stream()
+            .filter(
+                sentBlock ->
+                    combinedChainDataClient
+                        .getBlockAtSlotExact(sentBlock.getSlot())
+                        .join()
+                        .map(block -> block.equals(sentBlock))
+                        .orElse(false))
+            .count();
+
+    return new BlockPerformance((int) numberOfIncludedBlocks, producedBlocks.size());
+  }
+
+  private AttestationPerformance getAttestationPerformanceForEpoch(
+      UInt64 currentEpoch, UInt64 analyzedEpoch) {
+    checkArgument(
+        analyzedEpoch.isLessThanOrEqualTo(currentEpoch.minus(ATTESTATION_INCLUSION_RANGE)),
+        "Epoch to analyze attestation performance must be at least 2 epochs less than the current epoch");
+    // Attestations can be included in either the epoch they were produced in or in
+    // the following epoch. Thus, the most recent epoch for which we can evaluate attestation
+    // performance is current epoch - 2.
+    UInt64 analysisRangeEndEpoch = analyzedEpoch.plus(ATTESTATION_INCLUSION_RANGE);
+
+    // Get included attestations for the given epochs in a map from slot to attestations
+    // included in block.
+    Map<UInt64, List<Attestation>> attestationsIncludedOnChain =
+        getAttestationsIncludedInEpochs(analyzedEpoch, analysisRangeEndEpoch);
+
+    // Get sent attestations in range
+    Set<Attestation> producedAttestations =
+        producedAttestationsByEpoch.getOrDefault(analyzedEpoch, new HashSet<>());
+    BeaconState state = combinedChainDataClient.getBestState().orElseThrow();
+
+    int correctTargetCount = 0;
+    int correctHeadBlockCount = 0;
+    List<Integer> inclusionDistances = new ArrayList<>();
+
+    // Pre-process attestations included on chain to group them by
+    // data hash to inclusion slot to aggregation bitlist
+    Map<Bytes32, NavigableMap<UInt64, Bitlist>> slotAndBitlistsByAttestationDataHash =
+        new HashMap<>();
+    for (UInt64 slot : attestationsIncludedOnChain.keySet()) {
+      for (Attestation attestation : attestationsIncludedOnChain.get(slot)) {
+        Bytes32 attestationDataHash = attestation.getData().hash_tree_root();
+        NavigableMap<UInt64, Bitlist> slotToBitlists =
+            slotAndBitlistsByAttestationDataHash.computeIfAbsent(
+                attestationDataHash, __ -> new TreeMap<>());
+        Bitlist bitlistToInsert = attestation.getAggregation_bits().copy();
+        Bitlist alreadySetBits = slotToBitlists.get(slot);
+        if (alreadySetBits != null) {
+          bitlistToInsert.setAllBits(alreadySetBits);
+        }
+        slotToBitlists.put(slot, bitlistToInsert);
```

Comment:
(nit) Might be a bit simpler (and perform less copying) with something like:
```suggestion
        Bitlist bitlist = slotToBitlists.computeIfAbsent(slot, attestation.getAggregation_bits()::copy);
        bitlistToInsert.setAllBits(attestation.getAggregation_bits());
```

Reply:
For sure.